src/Database/Oracle/Simple/Internal.hs

```diff
@@ -818,44 +824,104 @@ stmtRelease
   -> IO ()
 stmtRelease = throwOracleError <=< dpiStmt_release
 
-data DPIData = DPIData
+-- | Used to write values to or read values from a column.
+data DPIData a = DPIData
   { dataIsNull :: CInt
-  , dataValue :: DPIDataBuffer
+  -- ^ If reading, a null value was read. If writing, writes a null value.
+  , dataValue :: a
+  -- ^ The value that was read/will be written, of type 'ReadBuffer' or 'WriteBuffer'.
   }
   deriving stock (Generic, Show, Eq)
   deriving anyclass (GStorable)
 
-newtype DPIDataBuffer = DPIDataBuffer (Ptr DPIDataBuffer)
+-- | An opaque pointer type for the @dpiDataBuffer@ union that we read from.
+-- We cannot write to this in a way that ODPIC could use.
+-- For poking purposes, use 'WriteBuffer'.
+newtype ReadBuffer = ReadBuffer (Ptr ReadBuffer)
   deriving (Show, Eq)
-  deriving newtype (Storable)
+  deriving newtype Storable
+
+-- | @dpiDataBuffer@ union that we can write to.
+-- We cannot read from this without a hint as to what type of data it contains.
+data WriteBuffer =
+  AsBoolean Int
+  | AsInt64 Int64
+  | AsUInt64 Word64
+  | AsFloat Float
+  | AsDouble Double
+  | AsString CString
+  | AsBytes DPIBytes
+  | AsTimestamp DPITimeStamp
+  deriving (Show, Eq, Generic)
+
+instance Storable WriteBuffer where
+  sizeOf _ = sizeOf (undefined :: DPITimeStamp) -- size of largest element
+  alignment _ = 8
+  peek = error "WriteBuffer says: don't peek in here!" -- handle this better!
```

Comment:
I'm okay with this as long as `WriteBuffer` is never exposed

Reply:
It will never see the light of day